pytensor/scalar/basic.py

```diff
@@ -3998,6 +3998,41 @@ class ScalarInnerGraphOp(ScalarOp, HasInnerGraph):
     def __init__(self, *args, **kwargs):
         self.prepare_node_called = set()
 
+    def _cleanup_graph(self, inputs, outputs):
+        # TODO: We could convert to TensorVariable, optimize graph,
+        # and then convert back to ScalarVariable.
+        # This would introduce rewrites like `log(1 + x) -> log1p`.
+
+        fgraph = FunctionGraph(copy(inputs), copy(outputs))
+
+        # Validate node types
+        for node in fgraph.apply_nodes:
+            if not isinstance(node.op, ScalarOp):
+                raise TypeError(
+                    f"The fgraph of {self.__class__.__name__} must be exclusively "
+                    "composed of scalar operations."
+                )
+
+        # Run MergeOptimization to avoid duplicated nodes
+        MergeOptimizer().rewrite(fgraph)
+
+        inputs, outputs = fgraph.inputs, fgraph.outputs
```

Comment:
You're overwriting the incoming parameters. Was that on purpose?

Asking in particular, because `old_output` below sounds like you were referring to the incoming parameter

Reply:
Yeah it's fine. That logic is just if we have `[out_A, out_B, out_A]`, we end up with `[out_A, out_B, clone(out_A)]`. There was an old bug when we didn't clone, and there is a direct test from when we fixed the bug